### PR TITLE
parser: fix HTTP version parsing

### PR DESCRIPTION
**Pull request by mscdex**
Only one digit is allowed for the major version and only one is allowed for the minor version [according to RFC 7230](https://tools.ietf.org/html/rfc7230#appendix-B).

### CODE REVIEW

File: test.c [hunk 1]
```diff
@@ -3313,9 +3313,9 @@ test_message_count_body (const struct message *message)
 }
 
 void
-test_simple (const char *buf, enum http_errno err_expected)
+test_simple_type (const char *buf, enum http_errno err_expected, enum http_parser_type type)
```
Comment (indutny) [inline test.c]: Nitpicking: does it violate 80 column limit?

File: http_parser.c [hunk 2]
```diff
@@ -1217,19 +1175,8 @@ size_t http_parser_execute (http_parser *parser,
 
         /* XXX allow spaces after digit? */
```
Comment (indutny) [inline http_parser.c]: Do we still want to allow any spaces after the digit?

File: http_parser.c [hunk 3]
```diff
@@ -795,75 +795,49 @@ size_t http_parser_execute (http_parser *parser,
...
-        }
-
-        parser->http_major *= 10;
-        parser->http_major += ch - '0';
-
-        if (UNLIKELY(parser->http_major > 999)) {
+        if (UNLIKELY(ch != '.')) {
           SET_ERRNO(HPE_INVALID_VERSION);
           goto error;
+          break;
```
Comment (indutny) [inline http_parser.c]: Superfluous break

File: http_parser.c [hunk 4]
```diff
@@ -795,75 +795,49 @@ size_t http_parser_execute (http_parser *parser,
...
         }
 
         parser->http_major = ch - '0';
-        UPDATE_STATE(s_res_http_major);
+        UPDATE_STATE(s_res_http_dot);
         break;
 
-      /* major HTTP version or dot */
-      case s_res_http_major:
+      case s_res_http_dot:
```
Comment (indutny) [inline http_parser.c]: I wonder if state machine for req and req can be united without sacrificing speed? Would branching in `http_minor` be too heavy for us?

File: http_parser.c (hunk 3 above)
Comment (mscdex) [inline http_parser.c]: Yeah I wasn't sure about it style-wise.

File: http_parser.c (hunk 4 above)
Comment (mscdex) [inline http_parser.c]: I dunno, I just went for the most straightforward change.

File: test.c (hunk 1 above)
Comment (mscdex) [inline test.c]: It is over 80 yes, but there are also many other lines in the file that are over 80 so I assumed that wasn't being observed, at least for that file...

Comment (indutny) [inline test.c]: We're moving to 80 column limit in a very very slow way :)

File: http_parser.c (hunk 2 above)
Comment (mscdex) [inline http_parser.c]: I would say no. It's pretty clear in the RFC that the HTTP version in the request line should be followed by CRLF.

File: http_parser.c (hunk 3 above)
Comment (mscdex) [inline http_parser.c]: Fixed.

File: http_parser.c (hunk 2 above)
Comment (indutny) [inline http_parser.c]: Let's remove this comment then.

File: test.c (hunk 1 above)
Comment (mscdex) [inline test.c]: Changed.

**Comment by indutny** [comment]
Landed in 9f489a4, thank you!

**Pull request closed without merging**